felix/statusrep/status_file_reporter.go

```diff
@@ -0,0 +1,305 @@
+				logrus.WithError(err).WithField("backoff", backoffDuration.String()).
+					Warn("Encountered an error during a retried update. Will retry again after a backoff...")
+
+				retryC = time.After(backoffDuration)
+			} else {
+				retryC = nil
+				fr.bom.reset()
+			}
+		case _, ok := <-scheduledResync:
+			if !ok {
+				logrus.Panic("Internal scheduled-resync channel closed unexpectedly")
+			}
+
+			err := fr.syncForeverReconcilePolicyFiles(true)
+			if err != nil {
+				backoffDuration := fr.bom.Step()
+				logrus.WithError(err).WithField("backoff", backoffDuration.String()).
+					Warn("Encountered an error during a scheduled re-sync. Queueing retry...")
+			}
+		}
+	}
+}
+
+// A sub-call of SyncForever, not intended to be called outside the main loop.
+// Updates delta tracker state to match the received update.
+// If commitToKernel is true, attempts to commit the new state to the kernel.
+// Can only return an error after a failed commit, so a returned error should
+// always result in SyncForever queueing a retry.
+func (fr *EndpointStatusFileReporter) syncForeverHandleEndpointUpdate(e interface{}, commitToKernel bool) error {
+	switch m := e.(type) {
+	case *proto.WorkloadEndpointStatusUpdate:
+		fr.policyDirDeltaTracker.Desired().Add(names.WorkloadEndpointIDToStatusFilename(m.Id))
+	case *proto.WorkloadEndpointStatusRemove:
+		fr.policyDirDeltaTracker.Desired().Delete(names.WorkloadEndpointIDToStatusFilename(m.Id))
+	default:
+		logrus.WithField("update", e).Warn("Skipping unrecognized endpoint update")
+		return nil
+	}
+
+	if commitToKernel {
+		err := fr.syncForeverReconcilePolicyFiles(false)
+		if err != nil {
+			return fmt.Errorf("Couldn't reconcile policy-status: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (fr *EndpointStatusFileReporter) syncForeverReconcilePolicyFiles(fullResync bool) error {
+	if fullResync {
+		// If calling this due to the first in-sync msg from upstream,
+		// this will be a no-op.
+		fr.policyDirDeltaTracker.Dataplane().DeleteAll()
+
+		// Load any existing committed dataplane entries.
+		entries, err := ensurePolicyStatusDir(fr.endpointStatusDirPrefix)
+		if err != nil {
+			return err
+		}
+		for _, entry := range entries {
+			fr.policyDirDeltaTracker.Dataplane().Add(entry.Name())
+		}
+	}
+
+	var lastError error
+	fr.policyDirDeltaTracker.PendingUpdates().Iter(func(name string) deltatracker.IterAction {
+		err := fr.writePolicyFile(name)
+		if err != nil {
+			logrus.WithError(err).Warn("Failed to write file to policy-status dir")
+			lastError = err
+			return deltatracker.IterActionNoOp
+		}
+
+		return deltatracker.IterActionUpdateDataplane
+	})
+
+	fr.policyDirDeltaTracker.PendingDeletions().Iter(func(name string) deltatracker.IterAction {
+		err := fr.deletePolicyFile(name)
+		if err != nil {
+			logrus.WithError(err).Warn("Failed to delete file in policy-status-dir")
+			// Carry on as normal (with a warning) if the file is somehow already deleted.
+			if !errors.Is(err, fs.ErrNotExist) {
```

Comment:
I think there's a standard helper `os.IsNotExist(err)`  not sure if this is equivalent

Reply:
Yep it's equivalent I believe, I went with this one after checking the docstring of `IsNotExist`:

   >  This function predates errors.Is. It only supports errors returned by the os
    package. New code should use errors.Is(err, fs.ErrNotExist).